main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -21,51 +21,670 @@ import ca.uwaterloo.flix.language.ast.{Ast, TokenErrorKind, ReadAst, SourceKind,
+   */
+  private def keyword(k: String)(implicit s: State): Boolean = {
+    // check if the keyword can appear before eof
+    if (s.current.offset + k.length > s.src.data.length) {
+      return false
+    }
+
+    val start = s.current.offset - 1
+    var matches = true
+    var offset = 0
+    while (matches && offset < k.length) {
+      if (s.src.data(start + offset) != k(offset)) {
+        matches = false
+      } else {
+        offset += 1
+      }
+    }
+    if (matches) {
+      for (_ <- 1 until k.length) {
+        advance()
+      }
+    }
+
+    matches
+  }
+
+  /**
+   * Moves current position past a built-in function, IE. "$BUILT_IN$".
+   */
+  private def builtIn()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (peek() == '$') {
+        advance()
+        return TokenKind.BuiltIn
+      }
+      advance()
+    }
+    TokenKind.Err(TokenErrorKind.UnterminatedBuiltIn)
+  }
+
+  /**
+   * Moves current position past all whitespace characters.
+   */
+  private def whitespace()(implicit s: State): Unit = {
+    while (!isAtEnd()) {
+      if (!peek().isWhitespace) {
+        return
+      }
+      advance()
+    }
+  }
+
+  /**
+   * Moves current position past a name (both upper- and lower-case).
+   * There are edge cases of variable holes, IE. "x?", and java names like "Map$Entry",
+   * which is the reason this function will return a `TokenKind`.
+   */
+  private def name(isUpper: Boolean)(implicit s: State): TokenKind = {
+    var kind = if (isUpper) {
+      TokenKind.NameUpperCase
+    } else {
+      TokenKind.NameLowerCase
+    }
+    while (!isAtEnd()) {
+      val p = peek()
+      if (p == '$' && peekPeek().exists(_.isLetter)) {
+        // Java names can have "$" in them. IE. "Map$Entry"
+        kind = TokenKind.NameJava
+      } else if (p == '?') {
+        advance()
+        return TokenKind.HoleVariable
+      } else if (!p.isLetter && !p.isDigit && p != '_' && p != '!') {
+        return kind
+      }
+      advance()
+    }
+    kind
+  }
+
+
+  /**
+   * Moves current position past a java name. IE. "##java"
+   */
+  private def javaName()(implicit s: State): TokenKind = {
+    advance()
+    while (!isAtEnd()) {
+      val p = peek()
+      if (!p.isLetter && !p.isDigit && p != '_' && p != '!') {
+        return TokenKind.NameJava
+      }
+      advance()
+    }
+    TokenKind.NameJava
+  }
+
+  /**
+   * Moves current position past a greek name.
+   * Greek names must lie in the unicode range U+0370 to U+03FF.
+   * IE. "Χαίρετε"
+   */
+  private def greekName()(implicit s: State): TokenKind = {
+    while (!isAtEnd()) {
+      if (!isGreekNameChar(peek())) {
+        return TokenKind.NameGreek
+      }
+      advance()
+    }
+    TokenKind.NameGreek
+  }
+
+  /**
+   * Checks whether `c` lies in unicode range U+0370 to U+03FF
+   */
+  private def isGreekNameChar(c: Char): Boolean = {
+    val i = c.toInt
+    i >= 880 && i <= 1023
```

Comment:
I would prefer the style `880 <= i && i <= 1023`.

Reply:
Any reason not to use hex decimal literals?

Reply:
Only that I hadn't thought about it! Will do :)